KAFKA-12368: Added inmemory implementations for RemoteStorageManager and RemoteLogMetadataManager.

satishd opened the pull request:
KAFKA-12368: Added inmemory implementations for RemoteStorageManager and RemoteLogMetadataManager.

* Added inmemory implementation for RemoteStorageManager and RemoteLogMetadataManager. A major part of inmemory RLMM will be used in the default RLMM implementation which will be based on topic storage. These will be used in unit tests for tiered storage.
* Added tests for both the implementations and their supported classes.

This is part of tiered storage implementation, [KIP-405](https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
KAFKA-12368: Added inmemory implementations for RemoteStorageManager and RemoteLogMetadataManager.

Added inmemory implementation for RemoteStorageManager and RemoteLogMetadataManager. A major part of inmemory RLMM will be used in the default RLMM implementation which will be based on topic storage. These will be used in unit tests for tiered storage.
Added a few tests for both the implementations.

This is part of tiered storage implementation, KIP-405.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,170 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.remote.storage;
```
It's a bit weird to include this in the client module. Since this is implemented in java, we could potentially create a new java module for it (like the Raft module). This reduces the size of the client jar and also avoids the inter-dependencies between java and scala.

Also, is this for testing? If so, it needs to be in the test directory.

satishd replied:
The plan was to use the related classes in the default RLMM implementation and move any class which is relevant only for tests to test dir later. I am +1 to have this as a separate module. I will update with those changes.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManager.java:
```diff
@@ -0,0 +1,169 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.remote.storage;
```
Is this for testing? If so, it needs to be in the test directory.

junrao reviewed (commented):
@satishd : Thanks for the PR. A couple of quick comments.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+import java.util.stream.Collectors;
+
+/**
+ * This class provides an imemory cache of remote log segment metadata. This maintains the lineage of segments
```
typo imemory

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+        idToSegmentMetadata.put(remoteLogSegmentId, updatedRlsm);
+        if (rlsmUpdate.state() != RemoteLogSegmentState.COPY_SEGMENT_STARTED) {
+            remoteLogSegmentIdInProgress.remove(remoteLogSegmentId);
+            addRemoteLogSegmentMetadata(updatedRlsm);
```
Hmm, it's possible for a segment to transition to DELETE_SEGMENT_STARTED here. Should those segments still be added?

satishd replied:
This behavior was kept to be the same as local log cleanup behavior, in which leader epoch is truncated only after local log is moved/deleted. Ideally, it is good not to consider the segments available that are being deleted as you said.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+            addRemoteLogSegmentMetadata(updatedRlsm);
+        }
+
+        if (rlsmUpdate.state() == RemoteLogSegmentState.DELETE_SEGMENT_FINISHED) {
```
Once a segment is in DELETE_SEGMENT_STARTED state, the corresponding segment could be gone any time after that. So, it seems that we should remove the segment from leaderEpochToOffsetToId once it's in DELETE_SEGMENT_STARTED?

satishd replied:
>This behavior was kept to be the same as local log cleanup behavior, in which leader epoch state is truncated only after local log is moved/deleted. Ideally, it is good not to consider the segments available that are being deleted as you said.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+
+    public Iterator<RemoteLogSegmentMetadata> listAllRemoteLogSegments() {
+        ArrayList<RemoteLogSegmentMetadata> list = new ArrayList<>(idToSegmentMetadata.values());
+        list.addAll(remoteLogSegmentIdInProgress.stream().map(id -> idToSegmentMetadata.get(id))
```
Could we just get the segment list from `idToSegmentMetadata.values()`?

satishd replied:
There may be few segments with state as `COPY_SEGMENT_STARTED` and they will be part of `remoteLogSegmentIdInProgress` only but not `idToSegmentMetadata`. That is why we need to add them to the list.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+                           : Collections.emptyIterator();
+    }
+
+    public Long highestLogOffset(int leaderEpoch) {
```
highestLogOffset => highestSegmentStartOffset ?

satishd replied:
No, it is not highestSegmentStartOffset but it is the highest log offset for the given leader epoch.
Nice catch! We need to give the max endoffset of all the segments for the given leader epoch.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,170 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class InmemoryRemoteLogMetadataManager implements RemoteLogMetadataManager {
```
Could we add a comment to this class?

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+        // look for floor entry as the given offset may exist in this entry.
+        Map.Entry<Long, RemoteLogSegmentId> entry = offsetToId.floorEntry(offset);
+        if (entry == null) {
+            // if the offset is lower than the minimum offset available in metadata then return null.
```
We are not returning null here.

junrao reviewed (commented):
@satishd : Thanks for the PR. A few more comments below.

satishd committed:
Moved inmemory implementations to a separate module.
Updated java doc comments and addressed typos.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+    private final ConcurrentMap<Integer, NavigableMap<Long, RemoteLogSegmentId>> leaderEpochToOffsetToId
+            = new ConcurrentHashMap<>();
+
+    public RemoteLogMetadataCache() {
```
This c'tor can be removed in exchange for the default generated c'tor.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    private void addRemoteLogSegmentMetadata(RemoteLogSegmentMetadata remoteLogSegmentMetadata) {
+        log.debug("Adding remote log segment metadata: [{}]", remoteLogSegmentMetadata);
```
It seems like we allow for an entry already existing with the same ID to be replaced with a different entry. Would that happen in practice?

satishd replied:
It may not occur in practice.

kowshik replied:
Is it useful to add a check against it?

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+
+    public Optional<RemoteLogSegmentMetadata> remoteLogSegmentMetadata(int leaderEpoch, long offset) {
+        NavigableMap<Long, RemoteLogSegmentId> offsetToId = leaderEpochToOffsetToId.get(leaderEpoch);
+        if (offsetToId == null || offsetToId.isEmpty()) {
```
Can `offsetToId` be empty if it is not null?
I understand it is right to check for emptiness here, but I was just curious to learn if it could happen in practice.

satishd replied:
This can happen in race condition when this method is queried while it was getting added in `addRemoteLogSegmentMetadata`. It may not happen in practice but it is good to have these checks.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+
+    private final Set<RemoteLogSegmentId> remoteLogSegmentIdInProgress = new HashSet<>();
+
+    private final ConcurrentMap<Integer, NavigableMap<Long, RemoteLogSegmentId>> leaderEpochToOffsetToId
```
Looking at the implementation, it appears we maintain some rules on when a `RemoteLogSegmentId` exists in one of these data structures versus all of them. It would be useful to briefly document those rules, and mention invariants (if any). For example, when an upload is in progress it is not (yet) added to this map.

satishd replied:
Sure, will add comments.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,152 @@
+    public void updateRemoteLogSegmentMetadata(RemoteLogSegmentMetadataUpdate rlsmUpdate)
+            throws RemoteResourceNotFoundException {
+        RemoteLogSegmentId remoteLogSegmentId = rlsmUpdate.remoteLogSegmentId();
+        RemoteLogSegmentMetadata rlsm = idToSegmentMetadata.get(remoteLogSegmentId);
```
can `metadata` be a better variable name over `rlsm`?

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,173 @@
+                    + " but it contains state as: " + remoteLogSegmentMetadata.state());
+        }
+
+        log.debug("Adding remote log segment : [{}]", remoteLogSegmentMetadata);
```
Should we move this log message before L51? so that the message conveying the intent is logged first before any possible  errors.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,173 @@
+
+        // Callers should use putRemoteLogSegmentMetadata to add RemoteLogSegmentMetadata with state as
+        // RemoteLogSegmentState.COPY_SEGMENT_STARTED.
+        if (rlsmUpdate.state() == RemoteLogSegmentState.COPY_SEGMENT_STARTED) {
```
Hmm... it seems like the only allowed state in `rlsmUpdate` is `COPY_SEGMENT_FINISHED`. Should we validate for that instead?

satishd replied:
It allows any state other than `COPY_SEGMENT_STARTED`, that is why we are checking only for this state.

satishd replied:
Other states include `COPY_SEGMENT_FINISHED`, `DELETE_SEGMENT_STARTED`, and `DELETE_SEGMENT_FINISHED`.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,173 @@
+
+        RemoteLogMetadataCache remoteLogMetadataCache = partitionToRemoteLogMetadataCache.get(topicIdPartition);
+        if (remoteLogMetadataCache == null) {
+            throw new RemoteResourceNotFoundException("No resource found for partition: " + topicIdPartition);
```
s/resource/entry ?

satishd replied:
"No <s>resource</s> metadata found for partition: "?

kowshik replied:
Sure

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,173 @@
+        idToPartitionDeleteMetadata.put(topicIdPartition, remotePartitionDeleteMetadata);
+        // there will be a trigger to receive delete partition marker and act on that to delete all the segments.
+
+        if (remotePartitionDeleteMetadata.state() == RemotePartitionDeleteState.DELETE_PARTITION_FINISHED) {
```
It seems like we want to add more protections here.

1. If `remotePartitionDeleteMetadata.state()` is `DELETE_PARTITION_FINISHED`, then should there have been a prior entry with `DELETE_PARTITION_STARTED` or  `DELETE_PARTITION_MARKED`?
2. Imagine there exists an entry in `partitionToRemoteLogMetadataCache` while the partition is also being deleted. Is that a valid state, or if not should we assert against it?

satishd replied:
1 -> Added more assertions. 
2 -> is a valid case.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,173 @@
+        log.debug("Adding delete state with: [{}]", remotePartitionDeleteMetadata);
+        TopicIdPartition topicIdPartition = remotePartitionDeleteMetadata.topicIdPartition();
+        idToPartitionDeleteMetadata.put(topicIdPartition, remotePartitionDeleteMetadata);
+        // there will be a trigger to receive delete partition marker and act on that to delete all the segments.
```
Sorry I don't understand what does this comment refer to?

satishd replied:
I meant there will be an external trigger based on delete partition marker, that is responsible for deleting the segments in a partition and updating the metadata. I will remove it as it looks to create confusion.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    @Override
+    public void onPartitionLeadershipChanges(Set<TopicIdPartition> leaderPartitions,
```
Hmm, any reason to not implement these methods? Is it that they don't serve any purpose in the in-memory implementation?

satishd replied:
Right, it is not applicable for inmemory implementation.

kowshik reviewed (commented):
@satishd Thanks for the PR! Few comments below.

satishd commented:
thanks @junrao @kowshik for the review comments, addressed with the latest commits/comments.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentState.java:
```diff
@@ -87,4 +89,27 @@ public byte id() {
+        Objects.requireNonNull(targetState, "targetState can not be null");
+
+        if (srcState == null) {
+            // If the source state is null, check the target state as the initial state viz DELETE_PARTITION_MARKED
```
DELETE_PARTITION_MARKED is not part of RemoteLogSegmentState.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentState.java:
```diff
@@ -87,4 +89,27 @@ public byte id() {
         return STATE_TYPES.get(id);
     }
+
+    public static boolean isValidTransition(RemoteLogSegmentState srcState, RemoteLogSegmentState targetState) {
```
This seems to be an internal implementation and is not part of the public API? Ditto for the same method in RemotePartitionDeleteState.

satishd replied:
This is not really an internal implementation but it validates the state transition and it is the same for any implementation.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+        idToSegmentMetadata.put(remoteLogSegmentId, updatedMetadata);
+        if (targetState != RemoteLogSegmentState.COPY_SEGMENT_STARTED) {
+            remoteLogSegmentIdInProgress.remove(remoteLogSegmentId);
+            addRemoteLogSegmentMetadata(updatedMetadata);
```
This method updates idToSegmentMetadata, which seems redundant since it's done in line 107 already.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+            for (Map.Entry<Integer, Long> entry : leaderEpochs.entrySet()) {
+                NavigableMap<Long, RemoteLogSegmentId> offsetToIds = leaderEpochToOffsetToId.get(entry.getKey());
+                // remove the mappings where this segment is deleted.
+                offsetToIds.values().remove(remoteLogSegmentId);
```
It's possible that after this, there is no segment associated with a leader epoch. Should we remove the entry with that leader epoch from leaderEpochToOffsetToId?

satishd replied:
Good point! It will clear the values which are empty maps.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        long max = 0L;
+        for (RemoteLogSegmentId id : offsetToSegmentId.values()) {
```
It's kind of inefficient to have to iterate through the whole segment list. Could we make leaderEpochToOffsetToId an ordered map and then do highEntry on that?

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+                           : Collections.emptyIterator();
+    }
+
+    public Optional<Long> highestLogOffset(int leaderEpoch) {
```
highestLogOffset => highestOffsetForEpoch?

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    @Override
+    public Iterator<RemoteLogSegmentMetadata> listRemoteLogSegments(TopicIdPartition topicIdPartition)
```
It seems that there is a semantic difference between this method and the next one. While this one exposes all segments (including in progress ones), the latter only exposes segments that are completed. It would be useful to document this clearly in the public API.

satishd replied:
I will update the javadoc of the APIs to make this clear.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+
+    public Iterator<RemoteLogSegmentMetadata> listAllRemoteLogSegments() {
+        ArrayList<RemoteLogSegmentMetadata> list = new ArrayList<>(idToSegmentMetadata.values());
+        list.addAll(remoteLogSegmentIdInProgress.stream().map(id -> idToSegmentMetadata.get(id))
```
> There may be few segments with state as COPY_SEGMENT_STARTED and they will be part of remoteLogSegmentIdInProgress only but not idToSegmentMetadata. That is why we need to add them to the list.

Hmm, it seems that we add the in-progress segment to idToSegmentMetadata in addToInProgress? It would be useful to add a comment for idToSegmentMetadata.

satishd replied:
Right, we do not need to add `remoteLogSegmentIdInProgress ` here.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+            addRemoteLogSegmentMetadata(updatedMetadata);
+        }
+
+        if (targetState == RemoteLogSegmentState.DELETE_SEGMENT_FINISHED) {
```
> This behavior was kept to be the same as local log cleanup behavior, in which leader epoch state is truncated only after local log is moved/deleted. Ideally, it is good not to consider the segments available that are being deleted as you said.

For the local log, we first schedule the segment for async deletion and then take it out of leaderEpochCache. So, the equivalent of that for remote storage seems to require taking the segment out of leaderEpochCache once the segment deletion is initiated.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+import java.util.stream.Collectors;
+
+/**
+ * This class provides an inmemory cache of remote log segment metadata. This maintains the lineage of segments
```
inmemory => in-memory

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments below.

kowshik commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentState.java:
```diff
@@ -87,4 +89,27 @@ public byte id() {
+            // Wanted to keep this logic simple here by taking null for srcState, instead of creating one more state like
+            // COPY_SEGMENT_NOT_STARTED and have the null check by caller and pass that state.
+            return targetState == COPY_SEGMENT_STARTED;
+        } else if (srcState == targetState) {
```
1. Will it be useful to place the implementation of this validation in a separate module, so that it can be reused with `RLMMWithTopicStorage` in the future?
2. Suggestion from the standpoint of code readability/efficiency: Would it make sense to replace the `if-else` logic by looking up from a `Map< RemoteLogSegmentState, Set< RemoteLogSegmentState>>` where key is the source state and value is a set of allowed target states?

satishd replied:
1 -> imho, this validation method should be part of the state enum and it can be used by any implementation including default RLMM. 

2 -> I would have preferred the suggested approach if there are many complex transitions but the transitions here are few and simple.

kowshik commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemotePartitionDeleteState.java:
```diff
@@ -83,4 +85,25 @@ public static RemotePartitionDeleteState forId(byte id) {
         return STATE_TYPES.get(id);
     }
 
+    public static boolean isValidTransition(RemotePartitionDeleteState srcState,
```
I have the same suggestions from `RemoteLogSegmentState` for this as well. Please refer to this comment: https://github.com/apache/kafka/pull/10218#discussion_r595492577

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,185 @@
+            new ConcurrentHashMap<>();
+
+    @Override
+    public void addRemoteLogSegmentMetadata(RemoteLogSegmentMetadata remoteLogSegmentMetadata)
```
Really minor comment/discussion: Any reason to call this prefixed with `add` as `addRemoteLogSegmentMetadata` vs calling the deletion one prefixed with `put` as `putRemotePartitionDeleteMetadata` i.e. instead can these 2 methods both start with the same prefix either `add` or `put`?

satishd replied:
`add` -> adding a new entry.
`put` -> add or update. 
`putRemotePartitionDeleteMetadata` is used for both add or update the `RemotePartitionDeleteMetadata`.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,185 @@
+/**
+ * This class is an implementation of {@link RemoteLogMetadataManager} backed by inmemory store.
+ */
+public class InmemoryRemoteLogMetadataManager implements RemoteLogMetadataManager {
```
We may want to think more about the locking semantics for this class and `RemoteLogMetadataCache`. 
Are we sure there would _not_ be use cases where we need to serialize mutations across the individually thread-safe attributes? If the answer is no, then using a fine-grained `Object` lock makes more sense because we can use it to guard critical sections.

Should we evaluate this upfront?

cc @junrao

satishd replied:
It was deliberate not to add locking semantics for now. We will add them once we have the respective changes using these classes.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,185 @@
+        Objects.requireNonNull(metadataUpdate, "metadataUpdate can not be null");
+
+        RemoteLogSegmentState targetState = metadataUpdate.state();
+        // Callers should use putRemoteLogSegmentMetadata to add RemoteLogSegmentMetadata with state as
```
In the comment:
s/putRemoteLogSegmentMetadata/addRemoteLogSegmentMetadata

kowshik reviewed (commented):
@satishd Thanks for the updated PR! Few more comments below.

satishd committed:
Extracted remote log leader epoch from cache into a separate class.
Refactoring for better modularization.
Addressed review comments.

kowshik commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentState.java:
```diff
@@ -21,14 +21,16 @@
 import java.util.stream.Collectors;
 
 /**
  * It indicates the state of the remote log segment. This will be based on the action executed on this
```
You can drop `It` and start with `Indicates the state...`.

kowshik commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemotePartitionDeleteState.java:
```diff
@@ -83,4 +85,25 @@ public static RemotePartitionDeleteState forId(byte id) {
+        if (srcState == null) {
+            // If the source state is null, check the target state as the initial state viz DELETE_PARTITION_MARKED
+            // Wanted to keep this logic simple here by taking null for srcState, instead of creating one more state like
+            // DELETE_PARTITION_NOT_MARKED and have the null check by caller and pass that state.
```
IMHO, we can simplify this to say:

```
// If the source state is null, check the target state as the initial state viz DELETE_PARTITION_MARKED.
// This ensures simplicity as we don't have to define an additional state type to represent the initial state.
```

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,185 @@
+    private final ConcurrentMap<TopicIdPartition, RemotePartitionDeleteMetadata> idToPartitionDeleteMetadata =
+            new ConcurrentHashMap<>();
+
+    private final ConcurrentMap<TopicIdPartition, RemoteLogMetadataCache> partitionToRemoteLogMetadataCache =
```
Could we call this `idToRemoteLogMetadataCache` to align with the naming of the other attribute thats called `idToPartitionDeleteMetadata` ?

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,185 @@
+
+        // this method is allowed only to add remote log segment with the initial state(which is RemoteLogSegmentState.COPY_SEGMENT_STARTED)
+        // but not to update the existing remote log segment metadata.
+        if (remoteLogSegmentMetadata.state() != RemoteLogSegmentState.COPY_SEGMENT_STARTED) {
```
Can this be checked inside `RemoteLogMetadataCache.addToInProgress()` instead of here?

kowshik commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentState.java:
```diff
@@ -87,4 +89,27 @@ public byte id() {
+    public static boolean isValidTransition(RemoteLogSegmentState srcState, RemoteLogSegmentState targetState) {
+        Objects.requireNonNull(targetState, "targetState can not be null");
+
+        if (srcState == null) {
```
It seems to me that `srcState` is never null in practice. Where does this check come into play in practice?

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+            = new ConcurrentHashMap<>();
+
+    // It keeps the segments which are not yet reached to COPY_SEGMENT_FINISHED state.
+    private final Set<RemoteLogSegmentId> remoteLogSegmentIdInProgress = new HashSet<>();
```
This is defined to be not thread safe unlike the other maps. Is there any reason?

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+public class RemoteLogMetadataCache {
+    private static final Logger log = LoggerFactory.getLogger(RemoteLogMetadataCache.class);
+
+    private final ConcurrentMap<RemoteLogSegmentId, RemoteLogSegmentMetadata> idToSegmentMetadata
```
Can we add a 1-line doc for this similar to other attributes below?

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+     *
+     * @param remoteLogSegmentMetadata
+     */
+    public void addToInProgress(RemoteLogSegmentMetadata remoteLogSegmentMetadata) {
```
Before we insert into the map/set, we should check if the provided `remoteLogSegmentMetadata.state()` is `COPY_SEGMENT_STARTED`.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+        if (!RemoteLogSegmentState.isValidTransition(existingMetadata.state(), targetState)) {
+            throw new IllegalStateException("Current state: " + existingState + ", target state: " + targetState);
+        }
+
```
In this method, we allow for existing entries in `idToSegmentMetadata` to be replaced, even if the state of the existing and new entries are the same. Is that intentional?

satishd replied:
Yes, it can happen to generate an event with the same state incase of retries.

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,172 @@
+        if (targetState == RemoteLogSegmentState.DELETE_SEGMENT_FINISHED) {
+            log.debug("Cleaning up the state for : [{}]", metadataUpdate);
+            // remove this entry when the state is moved to delete_segment_finished
+            Map<Integer, Long> leaderEpochs = existingMetadata.segmentLeaderEpochs();
```
Hmm, the entry for `existingMetadata` gets overwritten in the call to `addRemoteLogSegmentMetadata` in L110. Should we be accounting for the same here?

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,185 @@
+        RemoteLogSegmentState targetState = metadataUpdate.state();
+        // Callers should use putRemoteLogSegmentMetadata to add RemoteLogSegmentMetadata with state as
+        // RemoteLogSegmentState.COPY_SEGMENT_STARTED.
+        if (targetState == RemoteLogSegmentState.COPY_SEGMENT_STARTED) {
```
Similar to above comment, why not check this inside `remoteLogMetadataCache.updateRemoteLogSegmentMetadata()`?

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManagerTest.java:
```diff
@@ -0,0 +1,234 @@
+                System.currentTimeMillis(), SEG_SIZE, seg0leaderEpochs);
+        rlmm.addRemoteLogSegmentMetadata(segMetFooTp0s0e100);
+
+        // wWe should not get this as the segment is still gettign copied and it is not yet considered successful until
```
Typos:
1. s/wWe/We
2. s/gettign/getting

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManagerTest.java:
```diff
@@ -0,0 +1,234 @@
+    private static final int BROKER_ID = 0;
+
+    @Test
+    public void testRLMMFetchSegment() throws Exception {
```
Can we improve the local variable names? for example `segIdFooTp0s0e100`, `segMetFooTp0s0e100` etc. is not easy to read. We can use simpler names.

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCacheTest.java:
```diff
@@ -0,0 +1,149 @@
+                Arrays.asList(segCopyInProgress, segCopyFinished, segDeleteStarted));
+    }
+
+    private void checkContainsAll(Iterator<RemoteLogSegmentMetadata> allSegments,
```
The implementation compromises on the ordering, since it converts the iterator to a set. Is that intentional?

satishd replied:
Good catch. This is addressed in the latest commit.

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCacheTest.java:
```diff
@@ -0,0 +1,149 @@
+    private static final int BROKER_ID = 0;
+
+    @Test
+    public void testCacheSegmentsWithDifferentStates() throws Exception {
```
This particular test checks a number of things together in one test. Instead, could sections (1) to (4) from below each be defined as a separate unit test? Especially since each section seems to operate on a different segment, so it seems logically independent.

satishd replied:
I thought earlier about having different methods, but it checks `listAllSegments/listSegment(leaderEpoch)` APIs that return earlier segments. But I will have a separate test for that and extract as suggested.

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCacheTest.java:
```diff
@@ -0,0 +1,149 @@
+import java.util.Optional;
+import java.util.Set;
+
+public class RemoteLogMetadataCacheTest {
```
Could we add test(s) for `highestLogOffset` API?

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCacheTest.java:
```diff
@@ -0,0 +1,149 @@
+
+        // Search should return the above segment.
+        Optional<RemoteLogSegmentMetadata> seg3S350 = cache.remoteLogSegmentMetadata(0, 350);
+        Assertions.assertEquals(seg3.createRemoteLogSegmentWithUpdates(seg3Update1), seg3S350.orElse(null));
```
Could we assert just before this line that `seg3S350` is not empty? this will simplify the `seg3S350.orElse(null)` argument to `seg3S350.get()`.

(same comment applies for other places in this test)

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCacheTest.java:
```diff
@@ -0,0 +1,149 @@
+                System.currentTimeMillis(), SEG_SIZE, seg3leaderEpochs);
+        cache.addToInProgress(seg3);
+        RemoteLogSegmentMetadataUpdate seg3Update1 = new RemoteLogSegmentMetadataUpdate(seg3Id,
+                System.currentTimeMillis(), RemoteLogSegmentState.DELETE_SEGMENT_STARTED, BROKER_ID);
```
Should we alter the other arguments too, for example `BROKER_ID` and `eventTimestamp`? It appears that we expect `RemoteLogMetadataCache` to [apply all of the provided updates](https://github.com/apache/kafka/blob/0d9a95a7d0ab06aecc4480901707e29dd2a3147e/clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentMetadata.java#L240-L242), and this may include the other fields as well.

satishd replied:
Sure, I will add that.

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManager.java:
```diff
@@ -0,0 +1,171 @@
+    // map of key to log data, which can be segment or any of its indexes.
+    private Map<String, byte[]> keyToLogData = new ConcurrentHashMap<>();
+
+    public InmemoryRemoteStorageManager() {
```
Can we remove this c'tor in exchange for the default generated c'tor?

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManager.java:
```diff
@@ -0,0 +1,171 @@
+        Objects.requireNonNull(remoteLogSegmentMetadata, "remoteLogSegmentMetadata can not be null");
+        Objects.requireNonNull(logSegmentData, "logSegmentData can not be null");
+        try {
+            keyToLogData.put(generateKeyForSegment(remoteLogSegmentMetadata),
```
As per the interface we [mandate](https://github.com/apache/kafka/blob/0d9a95a7d0ab06aecc4480901707e29dd2a3147e/clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java#L76) the caller to ensure unique ID, but is it useful to add a guard that disallows replacing existing values?)

satishd replied:
We can add that.

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManager.java:
```diff
@@ -0,0 +1,171 @@
+            keyToLogData.put(generateKeyForIndex(remoteLogSegmentMetadata, IndexType.ProducerSnapshot),
+                    Files.readAllBytes(logSegmentData.producerSnapshotIndex().toPath()));
+        } catch (IOException e) {
+            throw new RemoteStorageException(e.getMessage(), e);
```
We could add a c'tor overload to [RemoteStorageException](https://github.com/apache/kafka/blob/0d9a95a7d0ab06aecc4480901707e29dd2a3147e/clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageException.java) that takes a `Throwable` as argument, it would the need to pass 2 args here.

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManager.java:
```diff
@@ -0,0 +1,171 @@
+        }
+
+        if (endPosition < startPosition) {
+            throw new IllegalArgumentException("end position must be greater than start position");
```
Probably better to say `...must be greater than or equal to...` ?

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManager.java:
```diff
@@ -0,0 +1,171 @@
+
+        // check for boundaries like given end position is more than the length, length should never be more than the
+        // existing segment size.
+        int length = Math.min(segment.length - 1, endPosition) - startPosition + 1;
```
Hmm, do we need to explicitly check if `endPosition` < `segment.length`?

satishd replied:
It takes `Math.min(endPosition, segment.length)`. So, no need to have that check.

kowshik commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManager.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
```
Is this intentionally left empty?

satishd replied:
Right.

kowshik reviewed (commented):
@satishd Thanks for the updates! Few more comments below, this time I made a pass on some of the test code as well.

satishd commented:
@kowshik I discussed the proposed changes in the call on 23rd,  and I mentioned that PR is not updated with those changes. I will let you know once those changes are pushed into this PR.

satishd commented:
@junrao @kowshik thanks for your review and comments. In the latest commit, I refactored the code with better abstractions and addressed the review comments.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,331 @@
+ * <li>
+ * {@link RemoteLogSegmentState#COPY_SEGMENT_STARTED}:
+ * <br>
+ * Segment in this state indicate it is not yet copied successfully. So, these segments will not be
```
indicate => indicates Ditto in a few other places.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,331 @@
+ * </li>
+ * </ul>
+ *
+ * <p>
```
The following table is a bit hard to read for developers. Since this is not meant for a public interface, could we make it more readable for developers?

satishd replied:
Sure, I will make it as simple ascii text.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,331 @@
+ * </p>
+ * <p></p>
+ */
+public class RemoteLogMetadataCache {
```
I guess this is an internal class. Will this be exposed in javadoc since currently it includes **/org/apache/kafka/server/log/remote/storage/* ?

satishd replied:
No. javadoc is generated for clients module with the package `/org/apache/kafka/server/log/remote/storage/ `. But this class is in `remote-storage` module.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,331 @@
+        // but not to update the existing remote log segment metadata.
+        if (remoteLogSegmentMetadata.state() != RemoteLogSegmentState.COPY_SEGMENT_STARTED) {
+            throw new IllegalArgumentException(
+                    "Given remoteLogSegmentMetadata should have state as " + RemoteLogSegmentState.COPY_SEGMENT_STARTED
```
Could we include remoteLogSegmentMetadata in the exception message?

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,331 @@
+            case COPY_SEGMENT_STARTED:
+                // Callers should use addCopyInProgressSegment to add RemoteLogSegmentMetadata with state as
+                // RemoteLogSegmentState.COPY_SEGMENT_STARTED.
+                throw new IllegalArgumentException("Metadata with " + RemoteLogSegmentState.COPY_SEGMENT_STARTED +
```
Should we include metadataUpdate in the message of the exception?

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,331 @@
+        NavigableMap<Integer, Long> epochToOffset = segmentMetadata.segmentLeaderEpochs();
+        Map.Entry<Integer, Long> nextEntry = epochToOffset.higherEntry(leaderEpoch);
+
+        // Update with the given metadata for leader epoch
```
This comment is confusing since there is no update here.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogLeaderEpochState.java:
```diff
@@ -0,0 +1,171 @@
+
+        // Add the old entry to unreferenced entries as the mapping is removed for the old entry.
+        if (oldEntry != null) {
+            unreferencedSegmentIds.add(oldEntry);
```
Hmm, during unclean leader election, some of the old segments may need to be added to unreferenced segment id list but may not have the exact offset of the new segment. How are those segments handled here?

satishd replied:
Yes, in the case of unclean leader election, the leader will remove the old segments for the respective leader epochs. The removal process involves removing the actual segment and updating the respective metadata of the segments.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,331 @@
+        long epochEndOffset = (nextEntry != null) ? nextEntry.getValue() - 1 : metadata.endOffset();
+
+        // Return empty when target offset > epoch's end offset or segment is not in COPY_SEGMENT_FINISHED state.
+        // This segment will not be available in offsetToId when it reaches the DELETE_SEGMENT_FINISHED state. So, no
```
It's weird to reference offsetToId here since it's in a separate class.

satishd replied:
Updated the comment.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,331 @@
+        // Return empty when target offset > epoch's end offset or segment is not in COPY_SEGMENT_FINISHED state.
+        // This segment will not be available in offsetToId when it reaches the DELETE_SEGMENT_FINISHED state. So, no
+        // need to add for that state here.
+        return (offset > epochEndOffset || metadata.state() != RemoteLogSegmentState.COPY_SEGMENT_FINISHED)
```
Is the test for RemoteLogSegmentState.COPY_SEGMENT_FINISHED necessary since it seems that only segments with RemoteLogSegmentState.COPY_SEGMENT_FINISHED exist in offsetToId.

satishd replied:
good point, this check is no more needed.

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogLeaderEpochState.java:
```diff
@@ -0,0 +1,171 @@
+        return highestLogOffset;
+    }
+
+    private void updateHighestLogOffset(long offset) {
```
updateHighestLogOffset => maybeUpdateHighestLogOffset ?

junrao commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,331 @@
+            = new ConcurrentHashMap<>();
+
+    // It contains leader epoch to the respective entry containing the state.
+    private final ConcurrentMap<Integer, RemoteLogLeaderEpochState> leaderEpochEntries = new ConcurrentHashMap<>();
```
When are entries in leaderEpochEntries removed?

satishd replied:
One way to do that is to clear the entry when the respective `RemoteLogLeaderEpochState` is empty. That means all the segments reached `DELETE_SEGMENT_FINISHED` state. 
This is not currently addressed. I plan to look into it when we integrate these APIs with RemoteLogManager by exploring other options too.

junrao commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,157 @@
+        RemotePartitionDeleteMetadata existingMetadata = idToPartitionDeleteMetadata.get(topicIdPartition);
+        RemotePartitionDeleteState existingState = existingMetadata != null ? existingMetadata.state() : null;
+        if (!RemotePartitionDeleteState.isValidTransition(existingState, targetState)) {
+            throw new IllegalStateException("Current state: " + existingState + ", target state: " + targetState);
```
Since existingState can be null, we want to handle it properly.

satishd replied:
It prints null. I may be missing something here. What needs to be handled here?

junrao commented on remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,157 @@
+    @Override
+    public Iterator<RemoteLogSegmentMetadata> listRemoteLogSegments(TopicIdPartition topicIdPartition)
+            throws RemoteStorageException {
+        return getRemoteLogMetadataCache(topicIdPartition).listAllRemoteLogSegments();
```
Should we requireNonNull for topicIdPartition here too?

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentState.java:
```diff
@@ -21,14 +21,16 @@
  * <p>
- * It goes through the below state transitions.
+ * It goes through the below state transitions. Self transition is treated as valid. This allows updating with the
+ * same state in case of retries and failover.
```
This is an existing issue. But is `&gt;` in line 37 expected?

satishd replied:
Looks like autoformatter changed it.

junrao commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageException.java:
```diff
@@ -28,9 +27,14 @@ public RemoteStorageException(final String message) {
         super(message, cause);
     }
 
+    public RemoteStorageException() {
```
Is this constructor needed?

satishd replied:
Not really needed for now.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments below.

satishd committed:
Addressed review comments.

satishd commented:
Thanks @junrao for the review.  Addressed them with the latest [commit](https://github.com/apache/kafka/pull/10218/commits/83050f3e4c03c5ac58c0ba2a9d8fcbf49661a1a9).

kowshik commented on clients/src/test/java/org/apache/kafka/test/TestUtils.java:
```diff
@@ -535,4 +536,48 @@ public static void setFieldValue(Object obj, String fieldName, Object value) thr
+     * @param iterator1 first iterator.
+     * @param iterator2 second iterator.
+     * @param <T>       type of element in the iterators.
+     * @return
```
nit: remove empty `@return`

kowshik commented on clients/src/test/java/org/apache/kafka/test/TestUtils.java:
```diff
@@ -535,4 +536,48 @@ public static void setFieldValue(Object obj, String fieldName, Object value) thr
+     * @param iterator1 first iterator.
+     * @param iterator2 second iterator.
+     * @param <T>       type of element in the iterators.
+     * @return
```
nit: remove empty `@return`

kowshik commented on remote-storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,305 @@
+            = new ConcurrentHashMap<>();
+
+    // It contains leader epoch to the respective entry containing the state.
+    private final ConcurrentMap<Integer, RemoteLogLeaderEpochState> leaderEpochEntries = new ConcurrentHashMap<>();
```
Should we call this map as `idToLeaderEpochState` or `idToEpochState` similar to the naming for the other map?

satishd replied:
The key is not really `id` but `epoch num`. What about `remoteLogLeaderEpochStateEntries` or `leaderEpochToState` or any other better name?

kowshik replied:
`leaderEpochToState` sounds good.

satishd committed:
Renamed 'remote-storage' module to 'storage' module.

satishd commented:
@junrao I renamed `remote-storage` module to `storage` module as you suggested in the commit [009d7fa](https://github.com/apache/kafka/pull/10218/commits/009d7fa6c861c671f3b2aa85c05fdd5ff467e195).

kowshik commented on clients/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentState.java:
```diff
@@ -87,4 +89,27 @@ public byte id() {
+    public static boolean isValidTransition(RemoteLogSegmentState srcState, RemoteLogSegmentState targetState) {
+        Objects.requireNonNull(targetState, "targetState can not be null");
+
+        if (srcState == null) {
```
Same comment as before: https://github.com/apache/kafka/pull/10218/files#r598982742.
Can srcState be null in practice? If not, this can be defined as an instance method.

satishd replied:
Yes, it can be null. It is called from [here](https://github.com/apache/kafka/pull/10218/files#diff-3724bb53d7ab4bc5a6ec4e1ab4c91c47bf90e4166d881f7706e2adc1848a5d16R293)

kowshik reviewed (commented):
@satishd Thanks for the update PR. Just few minor comments in this pass.

satishd commented:
@junrao This PR is on top of https://github.com/apache/kafka/pull/10489 . So, https://github.com/apache/kafka/pull/10489 be merged before merging this PR.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,160 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is an implementation of {@link RemoteLogMetadataManager} backed by in-memory store.
```
It would be useful to add a comment on whether the methods in this class are thread-safe or not.

satishd replied:
Sure, I will add the doc. They are currently not thread safe. But we want to address them when we integrate these APIs.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogLeaderEpochState.java:
```diff
@@ -0,0 +1,174 @@
+
+        // Update the highest offset entry for this leader epoch. This needs to be done as a segment can reach this
+        // state without going through COPY_SEGMENT_FINISHED state.
+        maybeUpdateHighestLogOffset(leaderEpochEndOffset);
```
It seems that it's inconsistent that we update highest log offset here but not in handleSegmentWithCopySegmentStartedState(). 

Could we comment on whether highestLogOffset reflects the segments that have reached  COPY_SEGMENT_FINISHED or not?

satishd replied:
After thinking through this more, we need to update this only when the segment reaches COPY_SEGMENT_FINISHED. This is effectively used to find out up to which offset the segments are already copied. I will remove the call here and keep the call only in handleSegmentWithCopySegmentFinishedState. WDYT?

junrao replied:
Sounds good. Could you make the change in the PR?

satishd replied:
Yes, I updated the PR.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,310 @@
+ * </li>
+ * </ul>
+ *
+ * <p>
```
It would be useful to document the meaning of the following table.

satishd replied:
Added a note.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManagerTest.java:
```diff
@@ -0,0 +1,130 @@
+        // Create remote log segment metadata and add them to RLMM.
+
+        // segment 0
+        // 0-100
```
Could we make it clear this is for offset range?

satishd replied:
Done

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteLogMetadataManagerTest.java:
```diff
@@ -0,0 +1,130 @@
+        Optional<RemoteLogSegmentMetadata> segMetadataForOffset30Epoch1 = rlmm.remoteLogSegmentMetadata(TP0, 1, 30L);
+        Assertions.assertEquals(Optional.of(expectedSegMetadata), segMetadataForOffset30Epoch1);
+
+        // Mark the partition for deletion. RLMM should clear all its internal state for that partition.
```
At this point, RLMM hasn't cleared all its internal state yet.

satishd replied:
Updated.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManagerTest.java:
```diff
@@ -0,0 +1,246 @@
+        for (Map.Entry<RemoteStorageManager.IndexType, Path> entry : expectedIndexToPaths.entrySet()) {
+            RemoteStorageManager.IndexType indexType = entry.getKey();
+            Path indexPath = entry.getValue();
+            log.info("Fetching index type: {}, indexPath: {}", indexType, indexPath);
```
Is this logging needed? Does it need to be in info level?

satishd replied:
We may have this as debug level by default. It will be helpful to see for which entry the test is failed.

junrao replied:
Could we move this to debug level then?

satishd replied:
Yes, it is done.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCacheTest.java:
```diff
@@ -0,0 +1,379 @@
+        // Create remote log segment metadata and add them to RemoteLogMetadataCache.
+
+        // segment 0
+        // 0-100
```
Could we make it clear this is for offset range?

satishd replied:
Updated.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCacheTest.java:
```diff
@@ -0,0 +1,379 @@
+            EpochOffset epochOffset = entry.getKey();
+            Optional<RemoteLogSegmentMetadata> segmentMetadata = cache.remoteLogSegmentMetadata(epochOffset.epoch, epochOffset.offset);
+            RemoteLogSegmentMetadata expectedSegmentMetadata = entry.getValue();
+            log.info("Searching for {} , result: {}, expected: {} ", epochOffset, segmentMetadata,
```
Is this logging needed? Ditto below.

satishd replied:
We may have this as debug level by default. It will be helpful to see for which `EpochOffset` the test is failed.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCacheTest.java:
```diff
@@ -0,0 +1,379 @@
+        //////////////////////////////////////////////////////////////////////////////////////////
+
+        Map<Integer, Long> expectedEpochToHighestOffset = new HashMap<>();
+        expectedEpochToHighestOffset.put(0, 19L);
```
It's kind of weird that the segment with epoch 0 is already deleted and yet we still expect the highest offset for epoch 0 to be returned.

satishd replied:
`highestLogOffset` can contain the deleted segments. `highestLogOffset` means the highest offset up to which the segments have been copied. Pl take a look at the [comment](https://github.com/apache/kafka/pull/10218#discussion_r609369253).

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCacheTest.java:
```diff
@@ -0,0 +1,379 @@
+        Assertions.assertTrue(TestUtils.sameElementsWithoutOrder(cache.listAllRemoteLogSegments(),
+                expectedSegmentsForEpoch0.iterator()));
+
+        // listRemoteLogSegments(0) should contain only segment2.
```
listRemoteLogSegments(0) => listRemoteLogSegments(1)

satishd replied:
Done

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments below.

satishd committed:
Addressed review comments

satishd commented:
@junrao Thanks for your comments. Replied on comments and addressed them with the [b5c0081](https://github.com/apache/kafka/pull/10218/commits/b5c0081b6f2740606e8458a3e784994ac5a5ee85) commit.

junrao commented on settings.gradle:
```diff
@@ -29,6 +29,7 @@ include 'clients',
     'log4j-appender',
     'metadata',
     'raft',
+    'storage',
```
This is redundant.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,309 @@
+            = new ConcurrentHashMap<>();
+
+    // It contains leader epoch to the respective entry containing the state.
+    private final ConcurrentMap<Integer, RemoteLogLeaderEpochState> leaderEpochEntries = new ConcurrentHashMap<>();
```
> One way to do that is to clear the entry when the respective RemoteLogLeaderEpochState is empty. That means all the segments reached DELETE_SEGMENT_FINISHED state.
> This is not currently addressed. I plan to look into it when we integrate these APIs with RemoteLogManager by exploring other options too.

Could we add a TODO comment here so that we don't forget about it?

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments.

kowshik commented on clients/src/test/java/org/apache/kafka/test/TestUtils.java:
```diff
@@ -535,4 +536,46 @@ public static void setFieldValue(Object obj, String fieldName, Object value) thr
+     * @param iterator2 second iterator.
+     * @param <T>       type of element in the iterators.
+     */
+    public static <T> boolean sameElementsWithOrder(Iterator<T> iterator1,
```
Here is a slightly simpler version:
```
 while (iterator1.hasNext() && iterator2.hasNext()) {
     if (!Objects.equals(iterator1.next(), iterator2.next())) {
            return false;
    }
}

return !iterator1.hasNext() && !iterator2.hasNext();
```

satishd replied:
IMHO, existing code looks easy to read/comprehend, and no multiple calls to hasNext().
How about the below code after removing inline variables in the existing code?

```
while (iterator1.hasNext()) {
    if (!iterator2.hasNext()) {
        return false;
    }

    if (!Objects.equals(iterator1.next(), iterator2.next())) {
        return false;
    }
}

return !iterator2.hasNext();
```

kowshik replied:
Sounds good

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogLeaderEpochState.java:
```diff
@@ -0,0 +1,172 @@
+
+        ArrayList<RemoteLogSegmentMetadata> metadataList = new ArrayList<>(size);
+        for (RemoteLogSegmentId id : offsetToId.values()) {
+            metadataList.add(idToSegmentMetadata.get(id));
```
Hmm here we assume that `id` should be present in the provided `idToSegmentMetadata`. Due to programming error, or other reasons, the caller may not be able to ensure this. Would it be safer if we instead threw whenever `id` is absent in `idToSegmentMetadata`  to catch that case?

satishd replied:
Good point, I will add a check for that.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogLeaderEpochState.java:
```diff
@@ -0,0 +1,172 @@
+
+    void handleSegmentWithCopySegmentStartedState(RemoteLogSegmentId remoteLogSegmentId) {
+        // Add this to unreferenced set of segments for the respective leader epoch.
+        unreferencedSegmentIds.add(remoteLogSegmentId);
```
The add call won't replace an existing element with the same `remoteLogSegmentId`. Is that expected?
For example, what happens if `addCopyInProgressSegment` is called twice but this line doesn't replace the existing entry?

satishd replied:
It already replaces the existing entry [here](https://github.com/apache/kafka/pull/10218/files#diff-3724bb53d7ab4bc5a6ec4e1ab4c91c47bf90e4166d881f7706e2adc1848a5d16R299).

kowshik replied:
Ok, I think this is fine then.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,309 @@
+                handleSegmentWithDeleteSegmentFinishedState(metadataUpdate, existingMetadata);
+                break;
+            default:
+                throw new IllegalArgumentException("Metadata with the state" + targetState + " is not supported");
```
nit: add one whitespace at the end after "...state"

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,309 @@
+     *
+     * @param remoteLogSegmentMetadata RemoteLogSegmentMetadata instance
+     */
+    public void addCopyInProgressSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata) {
```
Is this method expected to be idempotent?
Note: this comment is related to my other comment: https://github.com/apache/kafka/pull/10218#discussion_r610298531.

satishd replied:
Addressed in the above [comment](https://github.com/apache/kafka/pull/10218#discussion_r610462859).

kowshik replied:
Sounds good

kowshik reviewed (commented):
@satishd Thanks for the updated PR! Few more comments.

satishd committed:
Addressed review comments.

satishd commented:
Thanks @junrao @kowshik for the comments. Addressed them with the [d12c77a](https://github.com/apache/kafka/pull/10218/commits/d12c77a8f91d15cc9dd8b4c3f664f38c51c033b7).

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,316 @@
+            = new ConcurrentHashMap<>();
+
+    // It contains leader epoch to the respective entry containing the state.
+    // TODO We are not clearing the entry for epoch when RemoteLogLeaderEpochState becomes epty. This will be addressed
```
Typo epty

satishd replied:
Fixed.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. Just a minor comment below.

satishd committed:
Fixed a minor typo.

satishd commented:
Thanks @junrao for the comment, addressed with the commit [317be7a](https://github.com/apache/kafka/pull/10218/commits/317be7abe50264b6ed09bb85f36eacdc6b09599f).

junrao reviewed (approved):
@satishd : Thanks for the updated PR. LGTM. I will see if Kowshik has any further comments.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -0,0 +1,316 @@
+ *  The below table summarizes whether the segment with the respective state are available for the given methods.
+ * <pre>
+ * +---------------------------------+----------------------+------------------------+-------------------------+-------------------------+
+ * |  Method / SegmentState          | COPY_SEGMENT_STARTED | COPY_SEGMENT_FINISHED  | DELETE_SEGMENT_STARTED  | DELETE_SEGMENT_STARTED  |
```
typo: The title of the last column should be `DELETE_SEGMENT_FINISHED`.

satishd replied:
thanks, addressed it in the latest commit.

kowshik reviewed (approved):
@satishd: Thanks for the updated PR. Just a minor comment below. Let us wait for CI to pass before merge.

satishd committed:
Addressed a minor typo.

satishd commented:
@junrao @kowshik It looks like failures are not related to this PR.
